Tasks/AzureKeyVaultV1/operations/azure-arm-keyvault.ts

```diff
@@ -28,49 +31,85 @@ export class KeyVaultClient extends azureServiceClient.ServiceClient {
+                    if (!!vaultResourceId)
+                    {
                         console.log(tl.loc("RetryingWithVaultResourceIdFromResponse", vaultResourceId));
 
                         this.getCredentials().activeDirectoryResourceId = vaultResourceId; // update vault resource Id
                         this.getCredentials().getToken(true); // Refresh authorization token in cache
-                        var response = await this.beginRequest(request);
+                        response = await this.beginRequest(request);
                     }
                 }
 
-                if (retriableStatusCodes.indexOf(response.statusCode) != -1 && ++retryCount < maxRetryCount) {
-                    tl.debug(util.format("Encountered a retriable status code: %s. Message: '%s'.", response.statusCode, response.statusMessage));
-                    await webClient.sleepFor(timeToWait);
-                    timeToWait = timeToWait * retryIntervalInSeconds + retryIntervalInSeconds;
-                    continue;
+                if (this.retriableStatusCodes.indexOf(response.statusCode) === -1)
+                {
+                    return response;
                 }
 
-                return response;
-            } catch(error) {
-                if ((retriableErrorCodes.indexOf(error.code) != -1
-                || !!error.message && (error.message.startsWith('Request timeout: ') || error.message.startsWith('getaddrinfo ')))
-                && ++retryCount < maxRetryCount) {
-                    tl.debug(util.format("Encountered an error. Will retry. Error:%s. Message: %s.", error.code, error.message));
-                    await webClient.sleepFor(timeToWait);
-                    timeToWait = timeToWait * retryIntervalInSeconds + retryIntervalInSeconds;
+                if (++retryCount >= maxRetryCount)
+                {
+                    return response;
                 }
-                else {
+
+                tl.debug(`Encountered a retriable status code: ${response.statusCode}. Message: '${response.statusMessage}'.`);              
+            }
+            catch (error)
+            {
+                if (++retryCount >= maxRetryCount)
+                {
                     throw error;
                 }
+
+                if (!this.isRetriableError(error))
+                {
+                    throw error;
+                }
+
+                tl.debug(`Encountered an error. Will retry. Error: ${error.code}. Message: ${error.message}.`);                
             }
+
+            await webClient.sleepFor(retryIntervalInSeconds);
+        }
+    }
+
+    private isRetriableError(error: any): boolean
+    {
+        if (this.retriableErrorCodes.indexOf(error.code) !== -1)
+        {
+            return true;
         }
+
+        if (!error.message)
+        {
+            return false;
+        }
+
+        if (error.message.startsWith("Request timeout: "))
```

Comment:
I think you can combine all true statements:

`(error.message?.startsWith("Request timeout: ") || error.message?.startsWith("getaddrinfo "))`

It shouldn't enter the condition if `error.message` is `null` or `undefined`

And you can remove `!error.message` condition in that case.

Reply:
Unfortunately `?.` operator is available from Node 14 (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#browser_compatibility). However, I have introduced `retriableErrorMessages` array and check `error.message` against it. Thus the conditions here are simplified.